### PR TITLE
NEW: Feature Fixed Window Patch Extraction

### PR DESCRIPTION
**Pull request by shaneahmed**
- Rearrange FixedWindowPatchExtractor attributes for feature implementation.
- Addtional attribute `current_location` added to FixedWindowPatchExtractor .
- [x] `_find_optimal_level_and_downsample` is now public for accessibility in `patchextraction` module
- [x] `__next__` modified and added to base class
- [x] `__getitem__` modified and added to base class 
- [x] `_generate_location_df` added for window based patch extraction

### CODE REVIEW

File: tiatoolbox/tools/patchextraction.py [hunk 1]
```diff
@@ -53,29 +52,86 @@ class PatchExtractor(ABC):
         input_img(ndarray, WSIReader): input image for patch extraction.
           input_image type is ndarray for an image tile whereas :obj:`WSIReader`
           for an WSI.
-        patch_size(Tuple of int): patch size tuple (width, height).
-        resolution(Tuple of int): resolution at which to read the image.
+        patch_size(tuple(int)): patch size tuple (width, height).
+        resolution(tuple(int)): resolution at which to read the image.
```
Comment (vqdang) [inline tiatoolbox/tools/patchextraction.py:56]: Need to edit the doc as the code support both int and tuple(int)

File: tiatoolbox/tools/patchextraction.py [hunk 2]
```diff
@@ -53,29 +52,86 @@ class PatchExtractor(ABC):
...
+        locations_df(pd.DataFrame): A table containing location and/or type of patch.
 
     """
 
     def __init__(self, input_img, patch_size, resolution=0, units="level"):
-        self.patch_size = patch_size
+        if isinstance(patch_size, (tuple, list)):
+            self.patch_size = (int(patch_size[0]), int(patch_size[1]))
+        else:
+            self.patch_size = (int(patch_size), int(patch_size))
```
Comment (vqdang) [inline tiatoolbox/tools/patchextraction.py:67]: Highligh to refer back to above comment.

File: tiatoolbox/tools/patchextraction.py [hunk 3]
```diff
@@ -53,29 +52,86 @@ class PatchExtractor(ABC):
...
+        stride_w = self.stride[0]
+        stride_h = self.stride[1]
+
+        data = []
+
+        for h in range(int(math.ceil((img_h - img_patch_h) / stride_h + 1))):
+            for w in range(int(math.ceil((img_w - img_patch_w) / stride_w + 1))):
+                start_h = h * stride_h
+                start_w = w * stride_w
+                data.append([start_w, start_h, None])
```
Comment (vqdang) [inline tiatoolbox/tools/patchextraction.py:130]: You may want to use np.meshgrid to optimize this

File: tiatoolbox/tools/patchextraction.py [hunk 4]
```diff
@@ -107,18 +164,23 @@ def __init__(
...
             units=units,
         )
-        self.stride = stride
+        if stride is None:
+            self.stride = self.patch_size
+        else:
+            if isinstance(stride, (tuple, list)):
+                self.stride = (int(stride[0]), int(stride[1]))
+            else:
+                self.stride = (int(stride), int(stride))
```
Comment (vqdang) [inline tiatoolbox/tools/patchextraction.py:181]: Similar to above

File: tests/test_patch_extraction.py [hunk 5]
```diff
@@ -226,3 +227,112 @@ def test_points_patch_extractor_jp2(_sample_jp2):
...
+def test_fixed_window_patch_extractor():
+    """Test FixedWindowPatchExtractor for VF."""
+    file_parent_dir = pathlib.Path(__file__).parent
+    input_img = pathlib.Path(
+        file_parent_dir.joinpath("data/TCGA-HE-7130-01Z-00-DX1.png")
+    )
+
+    stride = (20, 20)
+    patch_size = (200, 200)
+    img = misc.imread(input_img)
```
Comment (vqdang) [inline tests/test_patch_extraction.py:241]: Do we need to test for not uniform stride and patch size?

File: tiatoolbox/tools/patchextraction.py (hunk 4 above)
Comment (vqdang) [reply tiatoolbox/tools/patchextraction.py:181]: Refer to the comment in test section, if we dont intend to support for not-uniform stride, then better to force input to be just int.

**Review by vqdang — APPROVED** [review]

**Review by mostafajahanifar — APPROVED** [review]

File: tiatoolbox/tools/patchextraction.py (hunk 1 above)
Comment (shaneahmed) [reply tiatoolbox/tools/patchextraction.py:56]: Fixed in this commit https://github.com/TIA-Lab/tiatoolbox/pull/97/commits/f85a61e9650b7dad8ffa73f6eafd37c5d18e5807

File: tiatoolbox/tools/patchextraction.py (hunk 2 above)
Comment (shaneahmed) [reply tiatoolbox/tools/patchextraction.py:67]: Fixed in this commit https://github.com/TIA-Lab/tiatoolbox/pull/97/commits/f85a61e9650b7dad8ffa73f6eafd37c5d18e5807

File: tiatoolbox/tools/patchextraction.py (hunk 4 above)
Comment (shaneahmed) [reply tiatoolbox/tools/patchextraction.py:181]: Fixed in these commits https://github.com/TIA-Lab/tiatoolbox/pull/97/commits/f85a61e9650b7dad8ffa73f6eafd37c5d18e5807 and https://github.com/TIA-Lab/tiatoolbox/pull/97/commits/d52f278d91a027a668335ab74b1a443bd76185ec

File: tests/test_patch_extraction.py (hunk 5 above)
Comment (shaneahmed) [reply tests/test_patch_extraction.py:241]: Fixed in this commit https://github.com/TIA-Lab/tiatoolbox/pull/97/commits/d52f278d91a027a668335ab74b1a443bd76185ec

File: tiatoolbox/tools/patchextraction.py (hunk 3 above)
Comment (shaneahmed) [reply tiatoolbox/tools/patchextraction.py:130]: I am leaving this for a separate PR as this needs to be carefully checked.

**Pull request merged**